File: src/samples/java/samples/TestLuaJit.java
package samples;

import com.sun.jna.Pointer;
import shibaInu.tolua.LuaState;
import shibaInu.tolua.ToLua;


/**
 * LUA / JIT / JAVA 运行效率测试
 * Created by LOLO on 2020/04/27.
 */
public class TestLuaJit {

    public void run() {

        LuaState state = new LuaState();
        state.addCallback("getTime", this::getTime);
        System.out.println();


        // jit 目前不支持在 MacOS 下运行
        if (!System.getProperty("os.name").startsWith("Mac OS")) {
            System.out.println("-----------[ JIT ]-----------");
            state.doFile("TestLuaJit_JitBytes");
            System.out.println();
        }


        // lua
        System.out.println("-----------[ LUA ]-----------");
        state.closeJIT();// 关闭 jit
        state.doFile("TestLuaJit");
        System.out.println();


        // java
        System.out.println("-----------[ JAVA ]----------");
        fn2();
        System.out.println();

    }

    // lua global function getTime()
    private int getTime(Pointer L) {
        ToLua.pushnumber(L, (int) System.currentTimeMillis());
        return 1;
    }


    // java

    private int[] arr1 = new int[999];
    private String[] arr2 = new String[]{"a", "b", "c", "d", "e"};

    private void fn1() {
        int n = 1;
        int len = arr1.length;
        int v1;
        String v2;
        for (int i = 0; i < 9999; i++) {
            if (i % 2 == 0) {
                n = n + 2;
                n = n * 2;
            } else {
                n = n - 2;
                n = n / 2;
            }

            for (int j = 0; j < len; j++) {
                v1 = arr1[j];
            }

            for (String v : arr2) {
                v2 = v;
            }
        }
    }

    private void fn2() {
        long t = 0;
        long time;
        for (int i = 0; i < 5; i++) {
            time = System.currentTimeMillis();
            fn1();
            t = System.currentTimeMillis() - time + t;
        }
        System.out.printf("5 times, ave: %sms%n", t / 5);
    }


    //
}

File: src/samples/lua/TestLuaJit.lua
--


local pairs = pairs
local arr1 = {}
local arr2 = { "a", "b", "c", "d", "e" }

local function fn1()
    local n = 1
    local len = 999
    local v1, v2
    for i = 1, 9999 do
        if i % 2 == 0 then
            n = n + 2
            n = n * 2
        else
            n = n - 2
            n = n / 2
        end

        for j = 1, len do
            v1 = arr1[j]
        end

        for k, v in pairs(arr2) do
            v2 = v
        end
    end
end


--
local t = 0

function fn2()
    local time
    for i = 1, 5 do
        time = getTime()
        fn1()
        t = getTime() - time + t
    end
    print(string.format("5 times, ave: %sms", t / 5))
end

fn2()


